libbeat/common/file/rotator.go

```diff
@@ -202,25 +218,134 @@ func (r *Rotator) Write(data []byte) (int, error) {
 		if err := r.openNew(); err != nil {
 			return 0, err
 		}
-	} else if r.intervalRotator != nil && r.intervalRotator.NewInterval() {
-		if err := r.rotate(rotateReasonTimeInterval); err != nil {
-			return 0, err
+	} else {
+		if reason, t := r.isRotationTriggered(dataLen); reason != rotateReasonNoRotate {
+			if err := r.closeFile(); err != nil {
+				return 0, errors.Wrap(err, "error file closing current file")
+			}
+
+			err := r.rot.Rotate(reason, t)
+			if err != nil {
+				return 0, err
+			}
+			if err = r.purge(); err != nil {
+				return 0, err
+			}
+
+			err = r.openFile()
+			if err != nil {
+				return 0, err
+			}
 		}
-		if err := r.openFile(); err != nil {
-			return 0, err
+	}
+
+	n, err := r.file.Write(data)
+	return n, errors.Wrap(err, "failed to write to file")
+}
+
+// openNew opens r's log file for the first time, creating it if it doesn't
+// exist.
+func (r *Rotator) openNew() error {
+	err := os.MkdirAll(r.dir(), r.dirMode())
+	if err != nil {
+		return errors.Wrap(err, "failed to make directories for new file")
+	}
+
+	_, err = os.Stat(r.rot.ActiveFile())
+	if err == nil {
+		// check if the file has to be rotated before writing to it
+		reason, t := r.isRotationTriggered(0)
+		if reason == rotateReasonNoRotate {
+			return r.appendToFile()
 		}
-	} else if r.size+dataLen > r.maxSizeBytes {
-		if err := r.rotate(rotateReasonFileSize); err != nil {
-			return 0, err
+		if err = r.rot.Rotate(reason, t); err != nil {
```

Comment:
again we can call `r.rotate` here, close is skipped and otherwise it's rotate + purge

Reply:
Again, `r.rotate` cannot accept the rotation time. But I have refactored it, so there is less duplication.